FhirSettings.getSettingsFilePath() evaluates user.home before looking at fhir.settings.path override and can cause startup failure

`org.hl7.fhir.utilities.settings.FhirSettings.getSettingsFilePath()` currently calls `getDefaultSettingsPath()` **before** checking the `fhir.settings.path` system property override.

Within this method it gets `user.home`, and then validates it. If it is `/`, it throws an exception and startup fails.

This breaks containerised deployments (e.g. OpenShift) where `user.home` may be `/ `
The `fhir.settings.path` override is never used due to the order of the code.

This is the code:
```
protected static String getSettingsFilePath(String explicitFilePath) {
    final String filePath;
    String pathFromSystemProperties;
    try {
      pathFromSystemProperties = getDefaultSettingsPath();
    } catch (IOException e) {
      pathFromSystemProperties = null;
    }
    filePath = explicitFilePath != null ? explicitFilePath :
      System.getProperty(FHIR_SETTINGS_PATH, pathFromSystemProperties);
    return filePath;
  }
```


**Suggested behavior**

Switch the order of the code above so that the explicitFilePath or FHIR_SETTINGS_PATH are used, and only compute the getDefaultSettingsPath if neither are set.


This is the full stacktrace:
```
Caused by: java.lang.RuntimeException: First entry in file path cannot be root: /, full path = /, .fhir, fhir-settings.json
	at org.hl7.fhir.utilities.PathBuilder.checkNonRootFirstEntry(PathBuilder.java:156)
	at org.hl7.fhir.utilities.PathBuilder.buildPath(PathBuilder.java:91)
	at org.hl7.fhir.utilities.Utilities.path(Utilities.java:550)
	at org.hl7.fhir.utilities.settings.FhirSettings.getDefaultSettingsPath(FhirSettings.java:219)
	at org.hl7.fhir.utilities.settings.FhirSettings.getSettingsFilePath(FhirSettings.java:196)
	at org.hl7.fhir.utilities.settings.FhirSettings.getInstanceFromPath(FhirSettings.java:188)
	at org.hl7.fhir.utilities.settings.FhirSettings.getInstance(FhirSettings.java:177)
	at org.hl7.fhir.utilities.settings.FhirSettings.isProhibitNetworkAccess(FhirSettings.java:136)
	at org.hl7.fhir.utilities.http.SimpleHTTPClient.get(SimpleHTTPClient.java:52)
	at org.hl7.fhir.utilities.http.ManagedWebAccessor.get(ManagedWebAccessor.java:107)
	at org.hl7.fhir.utilities.npm.PackageClient.fetchUrl(PackageClient.java:190)
	at org.hl7.fhir.utilities.npm.PackageClient.fetchCached(PackageClient.java:81)
	at org.hl7.fhir.utilities.npm.PackageClient.fetch(PackageClient.java:56)
	at org.hl7.fhir.utilities.npm.BasePackageCacheManager.loadFromPackageServer(BasePackageCacheManager.java:86)
	at ca.uhn.fhir.jpa.packages.loader.PackageLoaderSvc.fetchPackageFromServerInternal(PackageLoaderSvc.java:83)
	at ca.uhn.fhir.jpa.packages.loader.PackageLoaderSvc.fetchPackageFromPackageSpec(PackageLoaderSvc.java:78)
	at ca.uhn.fhir.jpa.packages.JpaPackageCache.loadPackageInner(JpaPackageCache.java:525)
	at ca.uhn.fhir.jpa.packages.JpaPackageCache.lambda$installPackage$3(JpaPackageCache.java:574)
	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:140)
	at ca.uhn.fhir.jpa.packages.JpaPackageCache.installPackage(JpaPackageCache.java:572)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:360)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:196)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:380)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:728)
	at ca.uhn.fhir.jpa.packages.JpaPackageCache$$SpringCGLIB$$0.installPackage(<generated>)
	at ca.uhn.fhir.jpa.packages.PackageInstallerSvcImpl.install(PackageInstallerSvcImpl.java:207)
	at ca.uhn.fhir.jpa.starter.common.StarterJpaConfig.packageInstaller(StarterJpaConfig.java:255)
	at ca.uhn.fhir.jpa.starter.common.StarterJpaConfig$$SpringCGLIB$$0.CGLIB$packageInstaller$3(<generated>)
	at ca.uhn.fhir.jpa.starter.common.StarterJpaConfig$$SpringCGLIB$$FastClass$$1.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:258)
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:400)
	at ca.uhn.fhir.jpa.starter.common.StarterJpaConfig$$SpringCGLIB$$0.packageInstaller(<generated>)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.lambda$instantiate$0(SimpleInstantiationStrategy.java:172)
	... 90 common frames omitted
```

To work around this issue, the following environment variable is required in OpenShift deployments:

```
env:
    - name: JAVA_TOOL_OPTIONS
       value: "-Duser.home=/tmp
```

--- a/org.hl7.fhir.utilities/src/main/java/org/hl7/fhir/utilities/settings/FhirSettings.java
+++ b/org.hl7.fhir.utilities/src/main/java/org/hl7/fhir/utilities/settings/FhirSettings.java
@@ -179,16 +179,23 @@ public class FhirSettings {
   }
 
   protected static String getSettingsFilePath(String explicitFilePath) {
-    final String filePath;
-    String pathFromSystemProperties;
-    try {
-      pathFromSystemProperties = getDefaultSettingsPath();
-    } catch (IOException e) {
-      pathFromSystemProperties = null;
+    if (explicitFilePath != null) {
+      log.debug("getting fhir-settings.json from explicit path: {}", explicitFilePath);
+      return explicitFilePath;
     }
-    filePath = explicitFilePath != null ? explicitFilePath :
-      System.getProperty(FHIR_SETTINGS_PATH, pathFromSystemProperties);
-    return filePath;
+
+    final String systemProperty = System.getProperty(FHIR_SETTINGS_PATH);
+    if (systemProperty != null) {
+      log.debug("getting fhir-settings.json from fhir.settings.path system property: {}", systemProperty);
+      return systemProperty;
+    }
+    try {
+      log.debug("getting fhir-settings.json from default location in user.home");
+      return getDefaultSettingsPath();
+    } catch (IOException e) {
+      log.error("Unable to construct default settings path from user.home", e);
+      return null;
+    }
   }
 
   static FhirSettingsPOJO getFhirSettingsPOJO(String filePath) throws IOException {

--- a/org.hl7.fhir.utilities/src/test/java/org/hl7/fhir/utilities/settings/FhirSettingsTests.java
+++ b/org.hl7.fhir.utilities/src/test/java/org/hl7/fhir/utilities/settings/FhirSettingsTests.java
@@ -16,23 +16,23 @@ import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.parallel.Isolated;
 
 @Isolated
-public class FhirSettingsTests implements ResourceLoaderTests {
+class FhirSettingsTests implements ResourceLoaderTests {
 
   private static String existingFhirSettingsPath;
 
   @BeforeAll
-  public static void beforeAll() {
+  static void beforeAll() {
     existingFhirSettingsPath = System.getProperty(FhirSettings.FHIR_SETTINGS_PATH);
     System.clearProperty(FhirSettings.FHIR_SETTINGS_PATH);
   }
 
   @AfterEach
-  public void afterEach() {
+  void afterEach() {
     System.clearProperty(FhirSettings.FHIR_SETTINGS_PATH);
   }
 
   @AfterAll
-  public static void afterAll() {
+  static void afterAll() {
     if (existingFhirSettingsPath == null) {
       System.clearProperty(FhirSettings.FHIR_SETTINGS_PATH);
     } else {
@@ -41,13 +41,13 @@ public class FhirSettingsTests implements ResourceLoaderTests {
   }
 
   @Test
-  public void testDefaultFhirSettingsPath() throws IOException {
+  void testDefaultFhirSettingsPath() throws IOException {
     String actualPath = FhirSettings.getSettingsFilePath(null);
     assertEquals(FhirSettings.getDefaultSettingsPath(), actualPath);
   }
 
   @Test
-  public void testJavaSystemPropertyFhirSettingsPath() throws IOException {
+  void testJavaSystemPropertyFhirSettingsPath() {
     final String dummyPath = "dummy-path";
     System.setProperty(FhirSettings.FHIR_SETTINGS_PATH, dummyPath);
     String actualPath = FhirSettings.getSettingsFilePath(null);
@@ -55,7 +55,7 @@ public class FhirSettingsTests implements ResourceLoaderTests {
   }
 
   @Test
-  public void testExplicitSettingsPathSelected() throws IOException {
+  void testExplicitSettingsPathSelected() {
     final String wrongDummyPath = "wrong-dummy-path";
     final String dummyPath = "dummy-path";
     System.setProperty(FhirSettings.FHIR_SETTINGS_PATH, wrongDummyPath);
@@ -64,21 +64,21 @@ public class FhirSettingsTests implements ResourceLoaderTests {
   }
 
   @Test
-  public void testExplicitFhirSettingsPath() throws IOException {
+  void testExplicitFhirSettingsPath() {
     final String dummyPath = "dummy-path";
     String actualPath = FhirSettings.getSettingsFilePath(dummyPath);
     assertEquals(dummyPath, actualPath);
   }
 
   @Test
-  public void testParseFhirSettings() throws IOException {
+  void testParseFhirSettings() throws IOException {
     Path path = Files.createTempFile("fhir-settings", "json").toAbsolutePath();
     copyResourceToFile(path, "settings", "settings-example.json");
 
     FhirSettingsPOJO fhirSettings = FhirSettings.getFhirSettingsPOJO(path.toString());
 
-   assertEquals("dummy-npm-path", fhirSettings.getNpmPath());
-   assertEquals("dummy-ruby-path", fhirSettings.getRubyPath());
+    assertEquals("dummy-npm-path", fhirSettings.getNpmPath());
+    assertEquals("dummy-ruby-path", fhirSettings.getRubyPath());
     assertEquals("dummy-api-key-value", fhirSettings.getApiKeys().get("dummy-api-key"));
     assertEquals("dummy-fhir-test-cases-path", fhirSettings.getFhirTestCasesPath());
     assertEquals("dummy-diff-tool-path", fhirSettings.getDiffToolPath());

